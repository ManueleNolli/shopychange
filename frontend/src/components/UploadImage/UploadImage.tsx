import {} from 'framer-motion'
import React, { useRef } from 'react'
import {
  FormControl,
  FormLabel,
  AspectRatio,
  InputGroup,
  Box,
  Image,
} from '@chakra-ui/react'

import { MdDelete } from 'react-icons/md'
import { Placeholder } from '../../assets/AssetsManager'

type UploadImageProps = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  onUpload: (event: any) => void
  cancelUpload: () => void
  currentImage: File | undefined
}

export default function UploadImage({
  onUpload,
  cancelUpload,
  currentImage,
}: UploadImageProps) {
  const inputRef = useRef<HTMLInputElement>(null)
  return (
    <FormControl isRequired>
      <FormLabel htmlFor="image">Image</FormLabel>
      <AspectRatio
        ratio={1}
        maxH={{ base: '40vh', md: '40vh' }}
        maxW={{ base: '40vh', md: '40vh' }}
        bg={'gray.100'}
        ml="auto"
        mr="auto"
      >
        {currentImage ? (
          <>
            <Image
              src={currentImage ? URL.createObjectURL(currentImage) : ''}
              alt="image"
            />
            <Box
              data-testid="delete-button"
              opacity={0}
              transition="opacity 0.3s"
              _groupHover={{ opacity: 0.5, cursor: 'pointer' }}
              onClick={cancelUpload}
            >
              <MdDelete fill="#FF4500" size={'50%'} />
            </Box>
          </>
        ) : (
          <InputGroup>
            <input
              data-testid="image-input"
              id="image"
              type="file"
              placeholder="image"
              accept="image/*"
              style={{ display: 'none' }}
              ref={inputRef}
              onChange={onUpload}
            />
            <Image
              data-testid="image-placeholder"
              src={Placeholder}
              w="100%"
              h="100%"
              onClick={() => {
                if (inputRef.current) {
                  inputRef.current.click()
                }
              }}
              _hover={{
                cursor: 'pointer',
                opacity: 0.5,
                transition: 'opacity 0.3s',
              }}
            />
          </InputGroup>
        )}
      </AspectRatio>
    </FormControl>
  )
}
