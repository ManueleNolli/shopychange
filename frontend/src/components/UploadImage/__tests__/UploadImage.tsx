import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import UploadImage from '../UploadImage'

const mockCreateObjectURL = jest.fn()
URL.createObjectURL = mockCreateObjectURL

describe('UploadImage', () => {
  test('renders correctly with currentImage', () => {
    const tree = render(
      <UploadImage
        onUpload={jest.fn()}
        cancelUpload={jest.fn()}
        currentImage={new File([''], 'filename', { type: 'image/png' })}
      />
    )
    expect(tree.container).toMatchSnapshot()
  })

  test('renders correctly without currentImage', () => {
    const tree = render(
      <UploadImage
        onUpload={jest.fn()}
        cancelUpload={jest.fn()}
        currentImage={undefined}
      />
    )
    expect(tree.container).toMatchSnapshot()
  })

  test('should call onUpload when input changes', () => {
    const onUpload = jest.fn()
    const tree = render(
      <UploadImage
        onUpload={onUpload}
        cancelUpload={jest.fn()}
        currentImage={undefined}
      />
    )
    const input = tree.getByTestId('image-input')
    const file = new File([''], 'filename', { type: 'image/png' })
    Object.defineProperty(input, 'files', {
      value: [file],
    })
    expect(onUpload).not.toHaveBeenCalled()

    fireEvent.change(input)

    expect(onUpload).toHaveBeenCalled()
  })

  test('should call cancelUpload when delete button is clicked', () => {
    const cancelUpload = jest.fn()
    const tree = render(
      <UploadImage
        onUpload={jest.fn()}
        cancelUpload={cancelUpload}
        currentImage={new File([''], 'filename', { type: 'image/png' })}
      />
    )

    const deleteButton = tree.getByTestId('delete-button')
    expect(cancelUpload).not.toHaveBeenCalled()

    fireEvent.click(deleteButton)

    expect(cancelUpload).toHaveBeenCalled()
  })

  test('should call input ref when image is clicked', () => {
    const tree = render(
      <UploadImage
        onUpload={jest.fn()}
        cancelUpload={jest.fn()}
        currentImage={undefined}
      />
    )

    const image = tree.getByTestId('image-placeholder')
    const input = tree.getByTestId('image-input')

    input.click = jest.fn()

    fireEvent.click(image)

    expect(input.click).toHaveBeenCalled()
  })
})
