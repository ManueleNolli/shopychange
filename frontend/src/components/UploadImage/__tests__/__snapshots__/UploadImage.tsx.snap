// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadImage renders correctly with currentImage 1`] = `
<div>
  <div
    class="chakra-form-control css-0"
    role="group"
  >
    <label
      class="chakra-form__label css-uvho5g"
      for="image"
      id="field-:r0:-label"
    >
      Image
      <span
        aria-hidden="true"
        class="chakra-form__required-indicator css-0"
        role="presentation"
      >
        *
      </span>
    </label>
    <div
      class="chakra-aspect-ratio css-oj014t"
    >
      <img
        alt="image"
        class="chakra-image css-0"
      />
      <div
        class="css-1lzlcxg"
        data-testid="delete-button"
      >
        <svg
          fill="#FF4500"
          height="50%"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="50%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`UploadImage renders correctly without currentImage 1`] = `
<div>
  <div
    class="chakra-form-control css-0"
    role="group"
  >
    <label
      class="chakra-form__label css-uvho5g"
      for="image"
      id="field-:r1:-label"
    >
      Image
      <span
        aria-hidden="true"
        class="chakra-form__required-indicator css-0"
        role="presentation"
      >
        *
      </span>
    </label>
    <div
      class="chakra-aspect-ratio css-oj014t"
    >
      <div
        class="chakra-input__group css-bx0blc"
        data-group="true"
      >
        <input
          accept="image/*"
          data-testid="image-input"
          id="image"
          placeholder="image"
          style="display: none;"
          type="file"
        />
        <img
          class="chakra-image css-glhh6s"
          data-testid="image-placeholder"
          src="placeholder.png"
        />
      </div>
    </div>
  </div>
</div>
`;
